diff --git a/lib-manual/overflow-15/overflow-15/overflow-sub-assign.rs b/lib-manual/overflow-15/overflow-15/overflow-sub-assign.rs
index 77f2096..1e36c84 100644
--- a/lib-manual/overflow-15/overflow-15/overflow-sub-assign.rs
+++ b/lib-manual/overflow-15/overflow-15/overflow-sub-assign.rs
@@ -1,7 +1,7 @@
 pub mod one;
 
 pub fn sub_assign(mut x: i32) -> i32 {
-    x -= one::one();
+    x.saturating_sub(one::one());
     
     x
 }
