diff --git a/lib-manual/dependency-1/dependency-1/dep.rs b/lib-manual/dependency-1/dependency-1/dep.rs
index 92726b0..9dea09b 100644
--- a/lib-manual/dependency-1/dependency-1/dep.rs
+++ b/lib-manual/dependency-1/dependency-1/dep.rs
@@ -4,5 +4,5 @@ pub fn dep() {
     let x: [i32; 2] = [0, 1];
     let y: i32 = x[1];
     let z: i32 = 2147483647;
-    let dep = z + y;
+    let dep = z.saturating_add(y);
 }
