diff --git a/lib-manual/overflow-9/overflow-9/overflow-shr.rs b/lib-manual/overflow-9/overflow-9/overflow-shr.rs
index 631ee5b..a063b9c 100644
--- a/lib-manual/overflow-9/overflow-9/overflow-shr.rs
+++ b/lib-manual/overflow-9/overflow-9/overflow-shr.rs
@@ -9,5 +9,5 @@ fn make_31() -> u32 {
 }
 
 pub fn shr(x: u32) -> u32 {
-    x >> make_31()
+    x.wrapping_shr(make_31())
 }
\ No newline at end of file
