diff --git a/lib-manual/overflow-8/overflow-8/overflow-shl-assign.rs b/lib-manual/overflow-8/overflow-8/overflow-shl-assign.rs
index d1b341d..0564fc1 100644
--- a/lib-manual/overflow-8/overflow-8/overflow-shl-assign.rs
+++ b/lib-manual/overflow-8/overflow-8/overflow-shl-assign.rs
@@ -9,7 +9,7 @@ fn make_31() -> u32 {
 }
 
 pub fn shl_assign(mut x: u32) -> u32 {
-    x <<= make_31();
+    x.wrapping_shl(make_31());
 
     x
 }
