diff --git a/lib-manual/overflow-17/overflow-17/overflow-shl.rs b/lib-manual/overflow-17/overflow-17/overflow-shl.rs
index aa43704..1171a69 100644
--- a/lib-manual/overflow-17/overflow-17/overflow-shl.rs
+++ b/lib-manual/overflow-17/overflow-17/overflow-shl.rs
@@ -1,5 +1,5 @@
 mod offset;
 
 pub fn shl(x: u32) -> u32 {
-    x << offset::make_31()
+    x.wrapping_shl(offset::make_31())
 }
\ No newline at end of file
