diff --git a/lib-manual/overflow-20/overflow-20/overflow-shr-assign.rs b/lib-manual/overflow-20/overflow-20/overflow-shr-assign.rs
index 7aa942f..9e8062e 100644
--- a/lib-manual/overflow-20/overflow-20/overflow-shr-assign.rs
+++ b/lib-manual/overflow-20/overflow-20/overflow-shr-assign.rs
@@ -1,7 +1,7 @@
 mod offset;
 
 pub fn shr_assign(mut x: u32) -> u32 {
-    x >>= offset::make_31();
+    x.wrapping_shr(offset::make_31());
 
     x
 }
