diff --git a/lib-manual/index-5/index-5/index.rs b/lib-manual/index-5/index-5/index.rs
index 25e96d0..b91cb6a 100644
--- a/lib-manual/index-5/index-5/index.rs
+++ b/lib-manual/index-5/index-5/index.rs
@@ -6,6 +6,6 @@ pub fn one() -> i32 {
 pub fn index() -> i32 {
     let x = [0];
     let mut y = 0;
-    let z = &x[one() as usize];
+    let z = &x[one() as usize - 1];
     return y;
 }
\ No newline at end of file
