From 33bbb5450d88bd9de02475f42679da8ba810f2e1 Mon Sep 17 00:00:00 2001
From: Josh Stone <cuviper@gmail.com>
Date: Tue, 17 Mar 2020 09:19:00 -0700
Subject: [PATCH] Fix CheckedDiv with zeros

---
 src/lib.rs | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/lib.rs b/src/lib.rs
index 634c20e..84f45a2 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -773,6 +773,9 @@ where
 {
     #[inline]
     fn checked_div(&self, rhs: &Ratio<T>) -> Option<Ratio<T>> {
+        if rhs.is_zero() {
+            return None;
+        }
         let gcd_ac = self.numer.gcd(&rhs.numer);
         let gcd_bd = self.denom.gcd(&rhs.denom);
         let denom = (self.denom.clone() / gcd_bd.clone())
@@ -1942,6 +1945,14 @@ mod test {
             assert_eq!(small.checked_div(&big), None);
             assert_eq!(_1.checked_div(&_0), None);
         }
+
+        #[test]
+        fn test_checked_zeros() {
+            assert_eq!(_0.checked_add(&_0), Some(_0));
+            assert_eq!(_0.checked_sub(&_0), Some(_0));
+            assert_eq!(_0.checked_mul(&_0), Some(_0));
+            assert_eq!(_0.checked_div(&_0), None);
+        }
     }
 
     #[test]
-- 
2.25.1

