diff --git a/lib-manual/overflow-7/overflow-7/overflow-shl.rs b/lib-manual/overflow-7/overflow-7/overflow-shl.rs
index e21defc..0ddb6e1 100644
--- a/lib-manual/overflow-7/overflow-7/overflow-shl.rs
+++ b/lib-manual/overflow-7/overflow-7/overflow-shl.rs
@@ -9,5 +9,5 @@ fn make_31() -> u32 {
 }
 
 pub fn shl(x: u32) -> u32 {
-    x << make_31()
+    x.wrapping_shl(make_31())
 }
\ No newline at end of file
