diff --git a/lib-manual/unwrap-1/unwrap-1/unwrap.rs b/lib-manual/unwrap-1/unwrap-1/unwrap.rs
index 23ada03..17855d1 100644
--- a/lib-manual/unwrap-1/unwrap-1/unwrap.rs
+++ b/lib-manual/unwrap-1/unwrap-1/unwrap.rs
@@ -1,5 +1,9 @@
 pub fn unwrap() {
     let x: Option<i32> = None;
-
-    x.unwrap();
+    match x {
+        Some(_) => {
+            x.unwrap();
+        }
+        _ => return,
+    }
 }
\ No newline at end of file
