diff --git a/lib-manual/overflow-6/overflow-6/overflow-mul-assign.rs b/lib-manual/overflow-6/overflow-6/overflow-mul-assign.rs
index 8aa6a33..692213d 100644
--- a/lib-manual/overflow-6/overflow-6/overflow-mul-assign.rs
+++ b/lib-manual/overflow-6/overflow-6/overflow-mul-assign.rs
@@ -1,5 +1,5 @@
 pub fn mul_assign(mut x: i32) -> i32 {
-    x *= 2;
+    x.saturating_mul(2);
     
     x
 }
\ No newline at end of file
