diff --git a/lib-manual/index-10/index-10/index.rs b/lib-manual/index-10/index-10/index.rs
index bfececa..4fa11e9 100644
--- a/lib-manual/index-10/index-10/index.rs
+++ b/lib-manual/index-10/index-10/index.rs
@@ -4,6 +4,9 @@ pub fn get_slice(x: Vec<i32>) -> Vec<i32> {
 
 pub fn index() {
     let x: [i32; 2] = [0, 1];
+    if 1 > 0 {
+        return;
+    }
     match get_slice(x[1..0].to_vec()) {
         _ => {}
     }
