diff --git a/lib-manual/unwrap-2/unwrap-2/unwrap.rs b/lib-manual/unwrap-2/unwrap-2/unwrap.rs
index c03f530..3e50277 100644
--- a/lib-manual/unwrap-2/unwrap-2/unwrap.rs
+++ b/lib-manual/unwrap-2/unwrap-2/unwrap.rs
@@ -1,6 +1,8 @@
 pub fn unwrap() {
     let x: Option<i32> = None;
     let mut y = 0;
-
-    y = x.unwrap();
+    y = match x {
+        Some(_) => x.unwrap(),
+        _ => return,
+    };
 }
\ No newline at end of file
